controllers/podchaos/containerkill/types.go

```diff
@@ -0,0 +1,190 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package containerkill
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"golang.org/x/sync/errgroup"
+
+	"github.com/go-logr/logr"
+
+	"github.com/pingcap/chaos-mesh/api/v1alpha1"
+	pb "github.com/pingcap/chaos-mesh/pkg/chaosdaemon/pb"
+	"github.com/pingcap/chaos-mesh/pkg/utils"
+
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	containerKillActionMsg = "delete container"
+)
+
+type Reconciler struct {
+	client.Client
+	Log logr.Logger
+}
+
+func (r *Reconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	var err error
+	now := time.Now()
+
+	r.Log.Info("reconciling container kill")
+	ctx := context.Background()
+
+	var podchaos v1alpha1.PodChaos
+	if err = r.Get(ctx, req.NamespacedName, &podchaos); err != nil {
+		r.Log.Error(err, "unable to get podchaos")
+		return ctrl.Result{}, nil
+	}
+
+	if podchaos.Spec.ContainerName == "" {
+		err = errors.New("the ContainerName is empty")
+		r.Log.Error(err, "the ContainerName is empty")
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	shouldAct := podchaos.GetNextStart().Before(now)
+	if !shouldAct {
+		return ctrl.Result{RequeueAfter: podchaos.GetNextStart().Sub(now)}, nil
+	}
+	pods, err := utils.SelectPods(ctx, r.Client, podchaos.Spec.Selector)
+	if err != nil {
+		r.Log.Error(err, "fail to get selected pods")
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	if len(pods) == 0 {
+		err = errors.New("no pod is selected")
+		r.Log.Error(err, "no pod is selected")
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	filteredPod, err := utils.GeneratePods(pods, podchaos.Spec.Mode, podchaos.Spec.Value)
+	if err != nil {
+		r.Log.Error(err, "fail to generate pods")
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	g := errgroup.Group{}
+	haveSelected := false
+	haveKilled := false
+	for index := range filteredPod {
+		pod := &filteredPod[index]
+
+		for index := range pod.Status.ContainerStatuses {
+			containerName := pod.Status.ContainerStatuses[index].Name
+			containerID := pod.Status.ContainerStatuses[index].ContainerID
+
+			if containerName == podchaos.Spec.ContainerName {
+				haveSelected = true
+				err = r.KillContainer(ctx, pod, containerID)
+				if err != nil {
+					r.Log.Error(err, "failed to kill container")
+				} else {
+					haveKilled = true
```

Comment:
If the `GeneratePods` function returns multiple pods, we should not use the execution results of one pod to represent all the results. we can remove `haveKilled` and `haveSelected`

Reply:
Maybe we can remove "haveKilled".  As for "haveSelected", if "haveSelected  == false", this means that there are no qualified containers in "pods". In this case, "haveSelected" needs to be retained.